### DISCUSSION TITLE
pair_allegro torch/tensor error in LAMMPs

### DISCUSSION THREAD

**Discussion by BryantLi-BLI** [mir-group/pair_nequip_allegro#37] [Q&A] [answered]
Hello! 

I've been trying to use pair_allegro with LAMMPs with my trained model `allegro-v3.pth`. The model itself is only trained on energy/force data (not including stress/virial). 

Pair_allegro is compiled with Libtorch-Cuda 12.1, and the pair_allegro is downloaded with the multicut branch. 

I tried to run a test case with LAMMPs, but it returned the following outputs. Does anyone know what may be the root cause of this? Also, it seems that this exact error repeats for every MPI process in this run.

```LAMMPS (2 Aug 2023)
  using 2 OpenMP thread(s) per MPI task

The 'box' command has been removed and will be ignored

Reading data file ...
  orthogonal box = (0 0 0) to (13.75725 13.75725 13.75725)
  2 by 4 by 4 MPI processor grid
  reading atoms ...
  128 atoms
  read_data CPU = 0.049 seconds
Allegro is using input precision f and output precision d
Exception: expected scalar type Double but found Float
Exception raised from check_type at aten/src/ATen/core/TensorMethods.cpp:16 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x6b (0x7f16c486e38b in /opt/libtorch/lib/libc10.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0xbf (0x7f16c4868f3f in /opt/libtorch/lib/libc10.so)
frame #2: <unknown function> + 0x2e3cf5f (0x7f171cc3cf5f in /opt/libtorch/lib/libtorch_cpu.so)
frame #3: double* at::TensorBase::mutable_data_ptr<double>() const + 0x30 (0x7f171cc3e020 in /opt/libtorch/lib/libtorch_cpu.so)
frame #4: at::TensorAccessor<double, 2ul, at::DefaultPtrTraits, long> at::TensorBase::accessor<double, 2ul>() const & + 0x5e (0x7f17312d169e in /opt/lammps/install/lib/liblammps.so.0)
frame #5: LAMMPS_NS::PairAllegro<(Precision)2>::compute(int, int) + 0x8cf (0x7f17312dc25f in /opt/lammps/install/lib/liblammps.so.0)
frame #6: LAMMPS_NS::Verlet::setup(int) + 0x3a2 (0x7f17313f6d22 in /opt/lammps/install/lib/liblammps.so.0)
frame #7: LAMMPS_NS::Run::command(int, char**) + 0xf69 (0x7f1731375a49 in /opt/lammps/install/lib/liblammps.so.0)
frame #8: LAMMPS_NS::Input::execute_command() + 0xb1d (0x7f17311c784d in /opt/lammps/install/lib/liblammps.so.0)
frame #9: LAMMPS_NS::Input::file() + 0x166 (0x7f17311c7c16 in /opt/lammps/install/lib/liblammps.so.0)
frame #10: <unknown function> + 0x13fc (0x55d3efc1d3fc in lmp)
frame #11: <unknown function> + 0x29d90 (0x7f16c4c01d90 in /lib/x86_64-linux-gnu/libc.so.6)
frame #12: __libc_start_main + 0x80 (0x7f16c4c01e40 in /lib/x86_64-linux-gnu/libc.so.6)
frame #13: <unknown function> + 0x1485 (0x55d3efc1d485 in lmp)

aborting job:
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0 
```

**Comment by Linux-cpp-lisp** [comment]
Hi @BryantLi-BLI ,

Looks like you have a trained model from an older version of Allegro/NequIP; so unless you're trying to run MD with NPT / use outputed pressures somehow, you can just use the older version of `pair_allegro`.

**Reply by BryantLi-BLI** [reply]
Hi, 

I want to use MD with NPT and this newer version of `pair_allegro`, so I'm aiming to just retrain my model. Assuming that, I tried to use `pip install nequip --upgrade` and installed the newest version of allegro from the GitHub page. Here are my current versions of nequip, allegro. However, after some testing it leads to the same error above. Is there a specific version/branch of nequip/allegro I need to use to generate the model for this version of `pair_allegro`?

NequIP Version: 0.5.6
mir-allegro Version: 0.2.0

**Reply by hmcezar** [reply]
Did you solve this issue @BryantLi-BLI ?

**Reply by Linux-cpp-lisp** [reply, marked as answer]
You'll both want the latest (current `develop` branch 0.6.0) of NequIP to avoid this error: https://github.com/mir-group/nequip/tree/develop
